library/Icingadb/Data/PivotTable.php

```diff
@@ -0,0 +1,410 @@
+                    $yAxisHeader => $table . '_' . $this->gridcols[$yAxisHeader]
+                ]);
+            } else {
+                $groupCols = $columns;
+            }
+
+            if ($groupCols !== null) {
+                // yAxisColumn and header might be the same column
+                $this->yAxisQuery->getSelectBase()->groupBy(array_values($groupCols));
+            }
+
+            $this->yAxisQuery->setColumns($columns);
+
+            if ($this->yAxisFilter !== null) {
+                $this->yAxisQuery->filter($this->yAxisFilter);
+            }
+
+            $this->yAxisQuery->orderBy(
+                $this->gridcols[$yAxisHeader]
+            );
+        }
+
+        return $this->yAxisQuery;
+    }
+
+    /**
+     * Return a pagination adapter for the x-axis query
+     *
+     * $limit and $page are taken from the current request if not given.
+     *
+     * @param   int     $limit  The maximum amount of entries to fetch
+     * @param   int     $page   The page to set as current one
+     *
+     * @return  Paginatable
+     */
+    public function paginateXAxis($limit = null, $page = null)
+    {
+        if ($limit === null || $page === null) {
+            if ($limit === null) {
+                $limit = $this->getPaginationParameter('x', 'limit', 20);
+            }
+
+            if ($page === null) {
+                $page = $this->getPaginationParameter('x', 'page', 1);
+            }
+        }
+
+        $query = $this->queryXAxis();
+
+        $query->limit($limit);
+
+        $query->offset($page > 0 ? ($page - 1) * $limit : 0);
+
+        return $query;
+    }
+
+    /**
+     * Return a Paginatable for the y-axis query
+     *
+     * $limit and $page are taken from the current request if not given.
+     *
+     * @param   int     $limit  The maximum amount of entries to fetch
+     * @param   int     $page   The page to set as current one
+     *
+     * @return  Paginatable
+     */
+    public function paginateYAxis($limit = null, $page = null)
+    {
+        if ($limit === null || $page === null) {
+            if ($limit === null) {
+                $limit = $this->getPaginationParameter('y', 'limit', 20);
+            }
+
+            if ($page === null) {
+                $page = $this->getPaginationParameter('y', 'page', 1);
+            }
+        }
+
+        $query = $this->queryYAxis();
+        $query->limit($limit);
+        $query->offset($page > 0 ? ($page - 1) * $limit : 0);
+
+        return $query;
+    }
+
+    /**
+     * Return the pivot table as an array of pivot data and pivot header
+     *
+     * @return array
+     */
+    public function toArray()
+    {
+        if (
+            ($this->xAxisFilter === null && $this->yAxisFilter === null)
+            || ($this->xAxisFilter !== null && $this->yAxisFilter !== null)
+        ) {
+            $xAxis = $this->getAxisKeys($this->queryXAxis(), $this->xAxisColumn);
```

Comment:
You should use PDO's native fetchPairs functionality here:

```php
$xAxis = $this->queryXAxis()->getDb()->fetchPairs($this->queryXAxis()->assembleSelect());
```

Do the same everywhere where you use `$this->getAxisKeys()`.